### PR TITLE
WIP: Support new file/upload file buttons on empty repo home page

### PR DESCRIPTION
**Pull request by lunny**
image

### CODE REVIEW

**Comment by xf-** [comment]
Maybe add the right portion of the normal bar on the top left side and set Download on disabled:
image

This would be a much more constant UI and user expierence

**Comment by lunny** [comment]
@xf- I don't think that's better than current design.

File: modules/repofiles/temp_repo.go [hunk 1]
```diff
@@ -54,22 +54,43 @@ func (t *TemporaryUploadRepository) Close() {
...
 func (t *TemporaryUploadRepository) Clone(branch string) error {
-	if _, stderr, err := process.GetManager().ExecTimeout(5*time.Minute,
-		fmt.Sprintf("Clone (git clone -s --bare): %s", t.basePath),
-		"git", "clone", "-s", "--bare", "-b", branch, t.repo.RepoPath(), t.basePath); err != nil {
-		if matched, _ := regexp.MatchString(".*Remote branch .* not found in upstream origin.*", stderr); matched {
-			return git.ErrBranchNotExist{
-				Name: branch,
+	if t.repo.IsEmpty {
+		if _, stderr, err := process.GetManager().ExecTimeout(5*time.Minute,
+			fmt.Sprintf("Clone (git clone -s --bare): %s", t.basePath),
```
Comment (techknowlogick) [inline modules/repofiles/temp_repo.go]: Can you move this command exec to git package?

File: modules/repofiles/temp_repo.go [hunk 2]
```diff
@@ -54,22 +54,43 @@ func (t *TemporaryUploadRepository) Close() {
...
-			return models.ErrRepoNotExist{
-				ID:        t.repo.ID,
-				UID:       t.repo.OwnerID,
-				OwnerName: t.repo.OwnerName,
-				Name:      t.repo.Name,
+		}
+	} else {
+		if _, stderr, err := process.GetManager().ExecTimeout(5*time.Minute,
+			fmt.Sprintf("Clone (git clone -s --bare): %s", t.basePath),
+			"git", "clone", "-s", "--bare", "-b", branch, t.repo.RepoPath(), t.basePath); err != nil {
```
Comment (techknowlogick) [inline modules/repofiles/temp_repo.go]: Same

File: modules/repofiles/temp_repo.go [hunk 3]
```diff
@@ -269,6 +293,18 @@ func (t *TemporaryUploadRepository) CommitTree(author, committer *models.User, t
 		"GIT_COMMITTER_EMAIL="+committerSig.Email,
 		"GIT_COMMITTER_DATE="+commitTimeStr,
 	)
+	if t.repo.IsEmpty {
+		commitHash, stderr, err := process.GetManager().ExecDirEnv(5*time.Minute,
+			t.basePath,
+			fmt.Sprintf("commitTree (git commit-tree): %s", t.basePath),
+			env,
+			"git", "commit-tree", treeHash, "-m", message)
```
Comment (techknowlogick) [inline modules/repofiles/temp_repo.go]: Same

**Comment by silverwind** [comment]
Think I'd also prefer having those button beside the clone URL. As an alternative, a bit less padding could also work. Or maybe even move the buttons into the header on the top-right.

**Comment by techknowlogick** [comment]
I like the idea of having the buttons next to clone, because that is what it is like when files are in the repo. So it would keep things the same for users.

If we don't have it next to clone, I like @silverwind's alternative of having less padding. 

A third alternative is to make the buttons Giant, so they fill up the empty whitespace.

Here is a screenshot of what GitHub does, but I don't like how they made the create/upload actions text:
Screen Shot 2019-05-16 at 3 27 52 PM

(sidenote: thanks @silverwind for GitHub dark theme)

**Comment by silverwind** [comment]
I like the idea of subtle text links. No need for big distracting buttons when most people will probably want to initizalize a repo through other means.

**Comment by mrsdizzie** [comment]
Agree it seems odd not to just have it in the same spot as every other page. Consistency across pages seems more important than anything else for this case

**Comment by lunny** [comment]
Should fix #4308

**Comment by delvh** [comment]
As of #14633, there is the request to include this ability also for the API.
Has that already been included in this branch?
Is that even the correct branch for that feature? I'd say so as it basically is concerned with the same topic, just from a slightly different perspective.
Or should a new branch be opened for that feature once this branch gets merged?

**Comment by lunny** [comment]
It will not be very difficult once this PR merged.

**Comment by jhult** [comment]
@lunny, any news on this PR? It would be a great feature to have.

**Comment by lunny** [comment]
> @lunny, any news on this PR? It would be a great feature to have.

Will continue this PR after some PRs merged.

**Comment by mikeperalta1** [comment]
Any update on this?

**Comment by lunny** [comment]
The code base has been changed a lot since this PR opened, just close it.

**Comment by ScarUY** [comment]
So, no way to add /create new file/s from the web ui? Honest question, I just installed it (rootless docker) and I'm stumped...

**Comment by lunny** [comment]
No, you could add /create new file/s from the web ui on a non-empty repository.
This PR is for empty repository.

**Comment by wxiaoguang** [comment]
Allow adding new files to an empty repo #24164 

All related issues could be re-reviewed and closed.

**Pull request closed without merging**